Adjust style checks to align with jwst

braingram opened the pull request:
This PR adjusts the style checks to roughly align with jwst (with a few adjustments for things like this package using cython).

For reference the ruff checks and pre-commit hooks for jwst are as follows:
https://github.com/spacetelescope/jwst/blob/e5c73fa23d157f4aa65416b4070c2b838e978494/pyproject.toml#L254
https://github.com/spacetelescope/jwst/blob/dd76184e82c1e863bce22d02de093f9bfe12f16f/.github/workflows/ci.yml

This PR sets the checks in this repo to roughly align:
- use default ruff rules
- extend ruff rules to include isort (which was previously a pre-commit hook in this repo)
- run bandit (like jwst does)
- enable cython-lint (as this package uses cython)

I applied auto-fixes in:
https://github.com/spacetelescope/stcal/pull/295/commits/a85bd00a2ceb5c9f1a26ea5992e04afd28c7741f
These are largely import sorting but I think it's worth a second set of eyes on these changes as I did notice one log message formatting change that likely indicates a bug in the log message (where it's currently not formatting the expected variable):
https://github.com/spacetelescope/stcal/pull/295/commits/a85bd00a2ceb5c9f1a26ea5992e04afd28c7741f#diff-c04284467dcbe86296aa89b83fd434e0dde8a32d790d157a9588ba3c7d169484R264

I then did some manual fixes removing unused variables and a duplicate test in:
https://github.com/spacetelescope/stcal/pull/295/commits/3c59de9a3fecca54438291c716a68e65e89939fa

There are still failures with this PR that I'm not sure how to address.
```

src/stcal/ramp_fitting/likely_algo_classes.py:289:24: F821 Undefined name `fit_ramps`
    |
287 |         n = alpha.shape[0]
288 |         z = np.zeros((len(cvec), len(countrates)))
289 |         result_low_a = fit_ramps(z, self, sig, countrateguess=countrates)
    |                        ^^^^^^^^^ F821
290 | 
291 |         # try to avoid problems with roundoff error
    |

src/stcal/ramp_fitting/likely_algo_classes.py:294:25: F821 Undefined name `fit_ramps`
    |
292 |         da_incr = da * (countrates[np.newaxis, :] + sig**2)
293 | 
294 |         result_high_a = fit_ramps(z, self, sig, countrateguess=countrates + da_incr)
    |                         ^^^^^^^^^ F821
295 |         # finite difference approximation to dw/da
    |

tests/test_ramp_fitting_likely_fit.py:613:34: F821 Undefined name `cube1`
    |
611 | def dbg_print_cube(cube, pix=(0, 0), label=None):
612 |     data, dq, vp, vr, err = cube
613 |     data1, dq1, vp1, vr1, err1 = cube1
    |                                  ^^^^^ F821
614 |     row, col = pix
615 |     nints = data1.shape[0]
```

2 are uses of an undefined `fit_ramps` in `likely_algo_classes.py`:
https://github.com/spacetelescope/stcal/blob/787aa81a6719bb0c72be8972171c7f3f88c8adf0/src/stcal/ramp_fitting/likely_algo_classes.py#L289
These would crash if anything tried to execute that code.

The last one is the use of an undefined `cube` in `dbg_print_cube` (which is unused). Calling this function would crash.

<!-- if you can't perform these tasks due to permissions, please ask a maintainer to do them -->
## Tasks
- [ ] update or add relevant tests
- [ ] update relevant docstrings and / or `docs/` page
- [ ] Does this PR change any API used downstream? (if not, label with `no-changelog-entry-needed`)
  - [ ] write news fragment(s) in `changes/`: `echo "changed something" > changes/<PR#>.<changetype>.rst` (see below for change types)
  - [ ] run regression tests with this branch installed (`"git+https://github.com/<fork>/stcal@<branch>"`)
    - [ ] [`jwst` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/jwst.yml) 
    - [ ] [`romancal` regression test](https://github.com/spacetelescope/RegressionTests/actions/workflows/romancal.yml)

<details><summary>news fragment change types...</summary>

- ``changes/<PR#>.apichange.rst``: change to public API
- ``changes/<PR#>.bugfix.rst``: fixes an issue
- ``changes/<PR#>.general.rst``: infrastructure or miscellaneous change
</details>

braingram commented:
@kmacdonald-stsci Any advice on how to handle the bugs revealed by these style checks?

Can `dbg_print_cube` be removed?

What is the expected definition of `fit_ramps` in `likely_algo_classes.py`?

kmacdonald-stsci commented:
> @kmacdonald-stsci Any advice on how to handle the bugs revealed by these style checks?
> 
> Can `dbg_print_cube` be removed?
> 
> What is the expected definition of `fit_ramps` in `likely_algo_classes.py`?

I have a few `dbg_` functions that I use for development and debugging, but aren't used at any other time.  I'd prefer not to remove those functions.  

The `fit_ramps` is defined here:

https://github.com/spacetelescope/stcal/blob/787aa81a6719bb0c72be8972171c7f3f88c8adf0/src/stcal/ramp_fitting/likely_fit.py#L493

braingram commented:
Thanks!

 > I have a few `dbg_` functions that I use for development and debugging, but aren't used at any other time. I'd prefer not to remove those functions.

I added a noqa for this line. It will however continue to fail if used.

> The `fit_ramps` is defined here:
> 
> https://github.com/spacetelescope/stcal/blob/787aa81a6719bb0c72be8972171c7f3f88c8adf0/src/stcal/ramp_fitting/likely_fit.py#L493

Since likely_fit imports likely_algo_classes:
https://github.com/spacetelescope/stcal/blob/787aa81a6719bb0c72be8972171c7f3f88c8adf0/src/stcal/ramp_fitting/likely_fit.py#L16
I think the import of `fit_ramps` needs to be local to avoid a circular import error. I updated the PR to include the local import.

braingram commented on src/stcal/ramp_fitting/ramp_fit.py:
```diff
@@ -261,7 +261,7 @@ def ramp_fit_data(
     # a minimum of a four group ramp is needed.
     ngroups = ramp_data.data.shape[1]
     if algorithm.upper() == "LIKELY" and ngroups < likely_fit.LIKELY_MIN_NGROUPS:
-        log.info(f"When selecting the LIKELY ramp fitting algorithm the"
```
Here's one likely bug in this log message. Since only first first item is a f-string all the subsequent items don't get formatted. This means the log message will contain the string "{likely_fit.LIKELY_MIN_NGROUPS}" instead of the formatted string for that variable.

braingram replied:
Thanks. Since it changed the log message I originally left it out of the style changes. I just pushed a fix in 84dbac6

melanieclarke replied:
Looks good, thanks for the fix!

braingram commented on tests/test_jump.py:
```diff
@@ -485,42 +484,6 @@ def test_find_faint_extended_sigclip():
     assert np.all(gdq[0, 4, 12:22, 14:23]) == 0
 
-# No shower is found because the event is identical in all ints
-def test_find_faint_extended_sigclip():
```
This test is duplicated. See:
https://github.com/spacetelescope/stcal/blob/787aa81a6719bb0c72be8972171c7f3f88c8adf0/tests/test_jump.py#L439
and
https://github.com/spacetelescope/stcal/blob/787aa81a6719bb0c72be8972171c7f3f88c8adf0/tests/test_jump.py#L489

braingram commented:
@kmacdonald-stsci any suggestions for a second reviewer? Most of the changes are code style but the duplicate jump test might be worth checking with a maintainer of that code to see if the duplicate is supposed to be different.

kmacdonald-stsci commented:
> @kmacdonald-stsci any suggestions for a second reviewer? Most of the changes are code style but the duplicate jump test might be worth checking with a maintainer of that code to see if the duplicate is supposed to be different.

For anything in STCAL, if you need an additional reviewer besides me, I think you can always tag @melanieclarke and/or @tapastro.

braingram commented:
Thanks for the quick review @melanieclarke Let me know how the jwst checks evolve and we can update these as well.

WilliamJamieson commented on .pre-commit-config.yaml:
```diff
@@ -15,49 +17,20 @@ repos:
-    rev: v0.15.0
+  - repo: https://github.com/PyCQA/bandit
+    rev: 1.7.10
     hooks:
-      - id: cython-lint
-      # - id: double-quote-cython-strings
+      - id: bandit
+        args: ["-r", "-ll", "src", "tests"]
```
Why is this being added? Ruff can do this check, https://docs.astral.sh/ruff/rules/#flake8-bandit-s.

braingram replied:
This was added to align the checks with jwst which uses bandit:
https://github.com/spacetelescope/jwst/blob/ffa02661593f4c53d66ab17fa837aa72a358def6/.pre-commit-config.yaml#L28

braingram replied:
No objection from me but that's beyond the scope of this PR.
